design/ceph-deprecate-old-osd-types.md

```diff
@@ -0,0 +1,433 @@
+aforementioned ConfigMap.
+
+As a matter of future security, the pod responsible for inspection should have read-only access to
+`/dev` (and `/run/udev` if needed). This stage should not have write access to any part of the host
+unless it is absolutely required.
+
+#### Stage 2: Process device inspection report (Rook-managed mode only)
+In the second stage, the operator examine the host device inspection report in the ConfigMap. The
+primary inspection processing done would be to filter the list of devices reported based on the
+`deviceFilter` cluster storage setting. This stage can also compare the host `devices` specified in
+the cluster config to the devices found during inspection.
+
+The insertion of stage 2 executed within the operator itself, placed in between stages 1 and 3
+allows for easy security improvements later, and it also helps keep the majority of the Ceph
+configuration logic in the operator itself rather than distributing the logic between multiple
+binaries. Existing logic distribution has contributed to operation which is not always transparent
+or easy to extend.
+
+#### Stage 3: Host device preparation (Rook-managed mode only)
+This stage is one in which `ceph-volume` provisions the devices. Unless a custom Rook binary is
+absolutely necessary, `ceph-volume` can be called directly from the pod's container entrypoint.
+Arguments to `ceph-volume` will be based on the host devices found in stage 1 and processing done
+in stage 2.
+
+For future security, this preparation pod should only need access to the individual disks which
+`ceph-volume` will use to provision Ceph.
+
+#### Stage 4: Host device (re)inspection (repeat stage 1)
+Repeat stage 1 inspection. For the **Ceph-managed mode**, this will be the first and only
+inspection stage. The `ceph-volume` report stored in the ConfigMap may now report existing
+OSDs on the host. These OSDs may have been created by Rook in Rook-managed mode, or they may have
+been created manually or via an action of Ceph's Rook orchestrator. The device report should contain
+the information necessary to determine which flags should be passed to `ceph-osd`.
+
+#### Stage 5: Start OSDs
+The last step remaining is to start the OSDs which were found by `ceph-volume`'s prior inspection.
+The device inspection report stored in step 4 will be used to determine which OSDs on the host
+belong to the operator's Ceph cluster (by UUID examination), and appropriate deployments will be
+created for each OSD.
+
+If possible, the existing trend of using init containers instead of a separate Rook binary
+(trend established for Ceph mons, mgrs, MDSes, and RGWs) should be used here with the intent being
+again to keep most of the logic in the operator. If teardown operations are needed, the Kubernetes
+container `PreStop` lifecycle hook should be used if reasonably possible.
+
+
+### Proposed Ceph mgr Rook orchestrator design
```

Comment:
Isn't this out of scope?

Reply:
I think this is really the core thing we are trying to achieve by v1.2 to integration with Octopus' operator. I think it's the complication of implementing this feature that results in the rest of the doc . The existing OSD code is hindering reasonable progress here, so a complete redesign is proposed.

Reply:
I've been discussing orchestrator things with Sebastian Wagner and Ricardo Dias this week, and I think the answer to your question is both yes and no. Parts of what I have written here should be part of an orchestrator design doc, but I think we do need to have some plans in place for how an OSD redesign in Rook can support the known orchestrator planning. I'm planning to write up an introductory orchestrator design to send to Sebastian and Ricardo and let them fill in more technical details, and I will have to modify this to be focused only on how Rook can support requests from the orchestrator.